File: backend/doppelganger/embeddings.py
"""Pluggable embeddings.

- LocalHashEmbedding: zero-dependency, deterministic. Good enough to prove
  the retrieval loop with no keys. NOT semantically strong — swap for prod.
- OpenAICompatEmbedding: any OpenAI-compatible /embeddings endpoint.
"""
from __future__ import annotations

import hashlib
import math
import re
from abc import ABC, abstractmethod

import httpx

from .config import settings

_TOKEN = re.compile(r"[a-z0-9']+")


class EmbeddingProvider(ABC):
    dim: int

    @abstractmethod
    async def embed(self, texts: list[str]) -> list[list[float]]: ...


class LocalHashEmbedding(EmbeddingProvider):
    """Hashing bag-of-words -> L2-normalized vector. Deterministic, offline."""

    def __init__(self, dim: int | None = None) -> None:
        self.dim = dim or settings.embedding_dim

    def _one(self, text: str) -> list[float]:
        vec = [0.0] * self.dim
        for tok in _TOKEN.findall(text.lower()):
            h = int(hashlib.md5(tok.encode()).hexdigest(), 16)
            vec[h % self.dim] += 1.0
        norm = math.sqrt(sum(v * v for v in vec)) or 1.0
        return [v / norm for v in vec]

    async def embed(self, texts: list[str]) -> list[list[float]]:
        return [self._one(t) for t in texts]


class OpenAICompatEmbedding(EmbeddingProvider):
    def __init__(self) -> None:
        self.dim = 0  # discovered from first response

    async def embed(self, texts: list[str]) -> list[list[float]]:
        headers = {"Authorization": f"Bearer {settings.embedding_api_key}"}
        async with httpx.AsyncClient(timeout=settings.request_timeout_seconds) as client:
            resp = await client.post(
                f"{settings.embedding_base_url}/embeddings",
                headers=headers,
                json={"model": settings.embedding_model, "input": texts},
            )
            resp.raise_for_status()
            vecs = [d["embedding"] for d in resp.json()["data"]]
            if vecs:
                self.dim = len(vecs[0])
            return vecs


def get_embedding_provider() -> EmbeddingProvider:
    if settings.embedding_provider == "openai":
        return OpenAICompatEmbedding()
    return LocalHashEmbedding()


def cosine(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    na = math.sqrt(sum(x * x for x in a)) or 1.0
    nb = math.sqrt(sum(y * y for y in b)) or 1.0
    return dot / (na * nb)

File: app/lib/engine/embeddings.dart
import 'dart:math';

/// Deterministic, dependency-free hash embedding — the on-device equivalent of
/// the backend's LocalHashEmbedding. Good enough to prove retrieval; swap for a
/// real embeddings API for quality.
class HashEmbedding {
  HashEmbedding({this.dim = 256});
  final int dim;

  static final RegExp _token = RegExp(r"[a-z0-9']+");

  // FNV-1a 32-bit: stable across runs, no crypto dependency.
  int _fnv1a(String s) {
    int hash = 0x811c9dc5;
    for (final c in s.codeUnits) {
      hash ^= c;
      hash = (hash * 0x01000193) & 0xFFFFFFFF;
    }
    return hash;
  }

  List<double> embed(String text) {
    final vec = List<double>.filled(dim, 0.0);
    for (final m in _token.allMatches(text.toLowerCase())) {
      vec[_fnv1a(m.group(0)!) % dim] += 1.0;
    }
    final norm = sqrt(vec.fold<double>(0.0, (a, v) => a + v * v));
    if (norm == 0) return vec;
    return vec.map((v) => v / norm).toList();
  }
}

double cosine(List<double> a, List<double> b) {
  double dot = 0, na = 0, nb = 0;
  for (var i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  if (na == 0 || nb == 0) return 0;
  return dot / (sqrt(na) * sqrt(nb));
}
